### ISSUE TITLE
尺度自适应自我注意力机制

### ISSUE THREAD

**Issue by suyu-star** [MCG-NJU/SparseBEV#79]
作者您好：
# Efficient implementation equivalent to the following:
    attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0) if is_causal else attn_mask
    attn_mask = attn_mask.masked_fill(not attn_mask, -float('inf')) if attn_mask.dtype==torch.bool else attn_mask
    attn_weight = torch.softmax((Q @ K.transpose(-2, -1) / math.sqrt(Q.size(-1))) + attn_mask, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p)
    return attn_weight @ V
根据torch源码，在计算attn weight 时attn_mask是与其直接相加的

    tau = tau.permute(0, 2, 1)
    attn_mask = dist[:, None, :, :] * tau[..., None]  # [B, 8, Q, Q]
    if pre_attn_mask is not None:  # for query denoising
        attn_mask[:, :, pre_attn_mask] = float('-inf')
    attn_mask = attn_mask.flatten(0, 1)  # [Bx8, Q, Q]
根据Sparse BEV源码，attn_mask对应$\tau D$

![image](https://github.com/MCG-NJU/SparseBEV/assets/67772212/754e5041-bbb2-4a13-b4b5-f4b4e884eabe)
但这和论文中的公式不符合，论文公式对应减法，上述的代码对应加法，请问是不是我哪里理解错了？

**Comment by afterthat97** [comment]
没问题啊，tau学成负数不就是等价的了

**Comment by FlyingPz** [comment]
他的dist是负数，因此后面也是负数
